src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -661,6 +642,24 @@ private void EnqueueWaiter(ConnectionWaiter waiter)
             _waitersHead = waiter;
         }
 
+        private bool HasWaiter()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObj));
+
+            if (_waitersHead != null)
+            {
+                Debug.Assert(_waitersTail != null);
+                Debug.Assert(Settings._maxConnectionsPerServer != int.MaxValue);
+                Debug.Assert(_idleConnections.Count == 0, $"With {_idleConnections} idle connections, we shouldn't have a waiter.");
+                return true;
+            }
+            else
+            {
+                Debug.Assert(_waitersTail == null);
+                return false;
+            }
```

Comment:
Very minor, but it looks like the JIT doesn't recognize that with optimizations this is just `return _waitersHead != null;` and insists on including a branch.  Might want to instead write:
```C#
private bool HasWaiter()
{
#if DEBUG
    … // whatever asserts here
#endif
    return _waitersHead != null;
}
```